3단계 - 테스트를 통한 코드 보호

jivebreaddev opened the pull request:
3단계 - 테스트를 통한 코드 보호

구현 요구사항
- 모든 bo에 대한 테스트 구현
[] OrderService
[] OrderTableService
[] ProductService 
[] MenuService 
[] MenuGroupService 

각각 mock과 fake로 구현했습니다.

UUIDInterface를 생성해서 의존도를 주입할 수 있게끔 변경하였습니다.

Jivebread committed:
chore(requirements): kitchenpos 요구사항 정리
- 1. erd diagram 추가
- 2. 구현사항들을 관찰해서 도메인에 대한 요구사항 정리

Jivebread committed:
chore(requirements): kitchenpos 요구사항 재정리
 1. NULL -> 비어있다 같은 비개발자적 표현
 2. 상태가 가지는 요구사항도 표기해주기
 3. 메뉴 카테고리 같은것들은 예시도 적어주면 이해하기 편하다.
 4. 0 보다 작을 수 없다. == 음수가 될 수 없다.
 5. 대기중 과 같은 상태에 대한 설명도 있으면 이해하기 편리하다.
 6. 매장 식사, 배달이면 필수인 정보들도 있으면 이해하기 편리하다.
 7. 메뉴 행동에서 누락된 부분 확인 필요하다.
 8. 메뉴의 가격이 올라가면 같은 요구사항이 틀린것으로 보이는 항목이 있다.

Jivebread committed:
chore(requirements): kitchenpos 요구사항 재정리
 1. 상태는 예시를 듦으로써 조금더 꼼꼼히 보여준다.
 2. 0보단 양수, 음수 사용
 3. `상품과 메뉴카테고리` 같은 오류 잡기
 4. 음식과 상품 같은 중복 명명 피하기
 5. 상태는 상태 시점을 정확히 표현하기
 6. BOOLEAN 값도 상태를 부여하여서 정확한 상태 지정하기
 7. 사용가능한 상태 -> 세세한 조건들 나열하기
 8. 중복제거
 9. 누락된 조건들 추가해서 작성하기

Jivebread committed:
test(Repository, Client): 가짜 객체를 주입하기위해 interface 모두 선언

Jivebread committed:
test(ProductFixture): ProductFixture 생성
- Product를 이름과, 가격이 주어졌을 때 생성

Jivebread committed:
test(UuidBuilder): Product에서 uuid(식별자)를 set하는 메서드를 제거하고 테스트를 쉽게 하기 위한 DI
- UuidBuilder 인터페이스 생성
- User에서 인터페이스 받아서 id 생성
- test에서 id 확인시에 해당 인터페이스로 확인.

Jivebread committed:
test(UuidBuilder): Product에서 uuid(식별자)를 set하는 메서드를 제거하고 테스트를 쉽게 하기 위한 DI
- UuidBuilder 인터페이스 생성
- User에서 인터페이스 받아서 id 생성
- test에서 id 확인시에 해당 인터페이스로 확인.

Jivebread committed:
test(ProductServiceTest): Product Mock 객체들을 사용해 테스트
- productServiceTest를 작성해서 요구사항 충족 확인

Jivebread committed:
test(FakeRepository, FakePurgomalumClient): 메뉴(Menu) 객체를 생성하기 위한 가짜 객체들 생성
- 가짜 객체 생성

Jivebread committed:
test(FakeUUID): 메뉴(UUIDBuilder)
FakeUuidBuilder를 생성하기 위해 INTERFACE 생성 및 주입

Jivebread committed:
test(MenuGroupServiceTest): 메뉴 서비스테스트
MenuGroupServiceTest를 요구사항에 대해서 진행

Jivebread committed:
test(Fake): Menu 테스트 진행을 위한 가짜 객체들 생성
MenuGroupServiceTest에서 service 주입을 위한 가짜객체들 생성

Jivebread committed:
test(MenuService): 메뉴 서비스 테스트
MenuServiceTest에서 요구사항들에 대한 검증 진행

Jivebread committed:
test(MenuService): 자리를 옮김에 따라 임시 commit

jivebreaddev committed:
test(OrderTableServiceTest): test OrderServiceTest with Fake Objects

jivebreaddev committed:
test(OrderServiceTest): test OrderServiceTest with Fake Objects

- Fixture들 생성
- FakeOrderTableRepository에서 잘못구현한 부분 수정했습니다.

jivebreaddev commented:
![image](https://github.com/next-step/ddd-legacy/assets/89639413/756d4312-1873-439c-b96e-8d2697d33e03)

Gomding commented on README.md:
````diff
@@ -6,13 +6,145 @@
 cd docker
 docker compose -p kitchenpos up -d
 ```
+## DB 엔티티 다이어그램
````
2단계 진행했던 부분도 커밋에 함께 들어갔군요 😢

Gomding commented on src/main/java/kitchenpos/infra/PurgomalumClient.java:
```diff
@@ -1,25 +1,5 @@
-            .toUri();
-        return Boolean.parseBoolean(restTemplate.getForObject(url, String.class));
-    }
+public interface PurgomalumClient {
```
현재 미션의 요구사항은 아니지만
만약 추상화하여 변경이 있다면
요구사항에 정리하신 `비속어` 라는 이름에 맞게 `PurgomalumClient` 란 이름을 변경해볼 수 있었을것 같네요 :)

`Purgomalum` 이란 이름은 비속어 검사해주는 외부 서비스이기 때문에 
interface 에 어울리는 이름은 아니라고 생각했어요 :)
키친포스 서비스에서 중요한건 어떤 외부서비스를 사용하는것이 아닌 `비속어를 검사한다` 이니까요 🤗

해당 코멘트는 반영하지 않으셔도 됩니다!!

jivebreaddev replied:
`BadWordsValidator`로 재명명하였습니다.

이런 디테일이 또 중요하잖습니까 하하하

챙겨주셔서 감사합니다!

Gomding commented on src/test/java/kitchenpos/fixture/MenuFixture.java:
```diff
@@ -0,0 +1,55 @@
+import java.math.BigDecimal;
+import java.util.List;
+
+public class MenuFixture {
```
테스트 객체 생성을 위한 Fixture 구현 👍 👍 👍 👍 👍 
앞으로 시창님이나 다른 개발자가 테스트를 편하게 작성할 수 있곘네요 :)
(퇴근 시간도 앞당겨지구요!)

jivebreaddev replied:
Fixture 구성을 하게되면 공통으로 사용하게되어 다른분들이 테스트를 작성하는데 부담이 확 줄어들것으로 보이네요!

Gomding commented on src/test/java/kitchenpos/infra/FakePurgomalumClient.java:
```diff
@@ -0,0 +1,11 @@
+package kitchenpos.infra;
+
+public class FakePurgomalumClient implements PurgomalumClient {
```
Fake 객체 구현 💯 💯 💯

jivebreaddev replied:
생성해두고 사용하지 않았네요;;

Gomding commented on src/test/java/kitchenpos/application/product/ProductServiceTest.java:
```diff
@@ -0,0 +1,141 @@
+  private MenuRepository menuRepository;
+
+  @Mock
+  private PurgomalumClient purgomalumClient;
```
여기는 `FakePurgomalumClient` 를 사용하지 않으신 이유가 궁금해요 :)
(MenuServiceTest, OrderServiceTest, OrderTableServiceTest 은 `FakePurgomalumClient` 사용하셔서요!)

jivebreaddev replied:
뚜렷한 이유가 없습니다.

누락입니다.

jivebreaddev replied:
제가 놓쳤네요 챙겨주셔서 감사합니다!

jivebreaddev replied:
ProductServiceTest는 Mock으로 진행해보고 싶었습니다.

확실히 구현에 대한 지식이 테스트를 작성할 때 필요해서 불편함을 체감하면서 진행했습니다.

Gomding replied:
다른 형태로 구현해보고 비교해보는것 좋네요 ㅎㅎ 👍

Gomding commented on src/test/java/kitchenpos/infra/FakePurgomalumClient.java:
```diff
@@ -0,0 +1,11 @@
+public class FakePurgomalumClient implements PurgomalumClient {
+  @Override
+  public boolean containsProfanity(String text) {
+    if (text.equals("badwords"))
```
아래와 같이 해도 동일하겠군요 🤗

```suggestion
    return text.equals("badwords");
```

jivebreaddev replied:
말씀하신 것과 같이 변경하였습니다

Gomding commented on src/test/java/kitchenpos/application/product/FakeUuidBuilder.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.HashMap;
+import java.util.UUID;
+
+public class FakeUuidBuilder implements UuidBuilder {
```
FakeUuidBuilder 요거는 product 패키지에 위치하지 않아도 될것같군요 :)
공용으로 쓰는것같아서요!

jivebreaddev replied:
domain > common > FakeUuidBuilder로 위치 변경하였습니다.

jivebreaddev replied:
UUID가 아닌 DB에 의해 생성된 DB ID의 경우에는 FakeRepository에서 ID를 1씩 증가시키시면서 진행하시나요?

Gomding replied:
넵 자동으로 증가하는 ID의 경우 1씩 증가하도록 구현할 수 있습니다 :)
`AtomicLong` 을 활용해볼 수 있죠 ㅎㅎ

```java
public class FakeSomethingRepository {
    private final AtomicLong id = new AtomicLong(0);
}
```

Gomding commented on src/test/java/kitchenpos/application/order/OrderServiceTest.java:
```diff
@@ -0,0 +1,275 @@
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class OrderServiceTest {
```
통합테스트를 하는 방법도 있었을텐데
가짜객체를 사용한 테스트를 진행하신 이유도 궁금해요 🙋‍♂️
가짜객체를 사용한 테스트에서 어떤 장점을 느끼셨을까요?

jivebreaddev replied:
### 느낀점 종합
1. 통합 테스트에 비해 빠르다!
2. 가짜객체로 원하는 상태와 행동을 지정하여서 테스트 할 수 있다.
3. 외부 infra 의존도가 가짜로 구현되서 테스트작성에 편하다. (실제 통합테스트가 아니라서 추가 테스트 작성필요함)

Gomding commented on src/main/java/kitchenpos/domain/OrderTableRepository.java:
```diff
@@ -1,8 +1,14 @@
 import java.util.UUID;
 
-public interface OrderTableRepository extends JpaRepository<OrderTable, UUID> {
+public interface OrderTableRepository {
```
유연한 구조로 만들어주셨군요 👍 
테스트에서 Fake 객체도 만들수 있구요 :)

Gomding commented on src/test/java/kitchenpos/application/menu/MenuGroupServiceTest.java:
```diff
@@ -0,0 +1,71 @@
+
+    @DisplayName("`메뉴 카테고리`의 `메뉴 카테고리 이름`은 비어있을 수 없다.")
+    @ParameterizedTest
+    @ValueSource(strings = {""})
```
`@EmptySource` 를 사용해볼 수도 있을것같아요!

null 검사까지 하려면

`@NullAndEmptySource` 가 있습니다 😁

jivebreaddev replied:
`@NullAndEmptySource` 를 사용해서 진행했습니다.

모르는 annotation 가르쳐주셔서 감사합니다! 요긴하게 사용하겠습니다!

Gomding commented on src/test/java/kitchenpos/application/menu/MenuGroupServiceTest.java:
```diff
@@ -0,0 +1,71 @@
+  @DisplayName("`메뉴 카테고리`(MENU CATEGORY)를 생성할 수 있다.")
+  class MenuCateogryRegistration {
+
+    @DisplayName("`메뉴 카테고리`의 `메뉴 카테고리 이름`은 비어있을 수 없다.")
```
요구사항에 있는것과 동일한 이름으로 테스트명을 작성해주셨네요 👍 👍 👍 
문서와 비슷하니 어떤 테스트인지 바로 알 수 있네요 ㅎㅎ

Gomding commented on src/test/java/kitchenpos/application/menu/MenuGroupServiceTest.java:
```diff
@@ -0,0 +1,71 @@
+              .isThrownBy(() -> menuGroupService.create(menuGroup));
+    }
+
+    @DisplayName("`메뉴 카테고리`의 `메뉴 카테고리 이름`을 생성할 수 있다.")
```
이 테스트는 
- `메뉴 카테고리` 생성 테스트일까요 !
- `메뉴 카테고리 이름`  생성 테스트일까요!

jivebreaddev replied:
`메뉴 카테고리 이름` 생성 테스트이네요;;

전형적인 service에 도메인 로직이 들어가있어서 헷갈리는 경우이네요!

Gomding replied:
상위 키워드는 메뉴 카테고리 생성인데
```java
  @Nested
  @DisplayName("`메뉴 카테고리`(MENU CATEGORY)를 생성할 수 있다.")
```

하위에는 갑자기 `메뉴 카테고리 이름 생성` 이 나와서 당황했어요!
해당 테스트는 `정상적인 이름일 때 메뉴 카테고리를 생성할 수 있다` 로 보이는데 다를까요?
`메뉴 카테고리 이름 생성` 이라는것은 `MENU CATEGORY NAME` 생성을 의미하는것같은데
MenuGroupService 에서 이름 생성이 여부가 중요한지 고민됐어요 ㅎㅎ

jivebreaddev replied:
1. 테스트 대상 Service 로직은 `메뉴 카테고리 생성`
2. 해당 테스트는 `MenuCategoryName 생성에 대한 테스트` 로 표기됨
3. `정상적인 이름일 때 메뉴 카테고리를 생성할 수 있다` 로 변경이 필요함

Gomding commented on src/test/java/kitchenpos/application/product/ProductServiceTest.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+
+    @DisplayName("상품의 이름(Product Name)이 없거나 비속어(Profanity)가 포함될 수 없다.")
```
이 테스트는 분리해보는것이 어떠신가요 :)
에러가 발생하는 상황이 다르다고 생각해서요 🤗
분리한다면 테스트가 실패했을 때 `빈값 검증 실패? 비속어 검사때문에 실패?` 이런 의문도 줄어들겠군요!

jivebreaddev replied:
ATDD를 진행하면서 `테스트갯수가 너무 많아지면 몰아넣어도 된다`고 하셔서 이렇게 진행했는데

두개의 테스트가 너무나 다른 테스트이었네요;;

확인해주셔서 감사합니다

Gomding commented on src/test/java/kitchenpos/application/product/ProductServiceTest.java:
```diff
@@ -0,0 +1,141 @@
+
+  }
+
+  @Nested
```
`@Nested` 를 사용하면 관련된 테스트를 묶을 수 있으니 가독성이 증가하겠네요 :)

Gomding reviewed (changes requested):
안녕하세요 시창님!
테스트 꼼꼼하게 작성해주셨군요 :)
가독성도 신경써주셨구요 👍 
몇 가지 코멘트 남겼으니 확인해주세요! 🙏 

궁금한 점 있으면 편하게 DM이나 코멘트 남겨주세요~

jivebreaddev committed:
feature(BadWordsValidator): 비속어검사 클라이언트 interface를 BadWordsValidator로 재명명하였습니다.

jivebreaddev committed:
feature(FakeBadWordsValidator): 비속어 검사 클라이언트 간결한 로직 작성

jivebreaddev committed:
feature(FakeUuidBuilder): UUID 생성하는 로직을 가짜객체로 구성하여서 test 가능하게 변경

jivebreaddev commented:
![image](https://github.com/next-step/ddd-legacy/assets/89639413/7e95f2a6-ae97-4ad3-b9a3-3a5f9a97f147)

## 요구사항

1. @EmptySource 사용해보기
2. client 클래스 재명명해보기
3. 부자연스러운 테스트케이스 두개로 분리하기

jivebreaddev committed:
test(TEST): @NullOrEmptySource 를 사용한 null 체크와 empty string 체크

jivebreaddev committed:
test(TEST): 너무나 다른 두가지의 테스트케이스 분리

jivebreaddev commented:
마지막으로 한가지 질문이 있습니다!

![image](https://github.com/next-step/ddd-legacy/assets/89639413/11ef0bcc-3e33-47e2-a614-5602305a50a0)

이렇게 많은 의존성을 추가하고 
다른 menu나 product repository에 상태값을 추가하는 경우 제가 잘하고 있는지 궁금합니다.

Gomding commented on src/test/java/kitchenpos/application/menu/MenuServiceTest.java:
```diff
@@ -0,0 +1,201 @@
+  public static final String FOR_TWO = "이인용";
+  public static final String TWO_UDONS = "우동과라면";
+
+  private MenuGroupRepository menuGroupRepository;
```
> 이렇게 많은 의존성을 추가하고
> 다른 menu나 product repository에 상태값을 추가하는 경우 제가 잘하고 있는지 궁금합니다.

제가 질문을 잘 이해했는지 모르겠지만
menu 나 product 에 수정이 있을 때 테스트 코드도 변경사항이 많이 생기는것이 걱정되는 것일까요?
테스트코드도 결국 유지보수 대상이라 생각해요 :)

상태값이 추가되는 경우 Fixture 를 통해 조금은 제어할 수 있을것 같아요
현재는 거의 모든 필드를 주입받는 Fixture 메서드들이 존재하는데
특정 값만 인자로 받고 나머지는 기본값으로 처리해준다면 해당 상태값이 필요없는 테스트들은 수정을 안할수도 있어요

아래와 같이요

```java
public class MenuFixture {
    private Menu menu;

    private MenuTestFixture(Menu menu) {
        this.menu = menu;
    }

    public static MenuTestFixture create() {
        Menu menu = new Menu();
        menu.setId(UUID.randomUUID());
        menu.setName("테스트 메뉴");
        menu.setPrice(BigDecimal.valueOf(10000));
        menu.setMenuGroup(MenuGroupTestFixture.create().getMenuGroup());
        menu.setMenuGroupId(UUID.randomUUID());
        menu.setDisplayed(true);
        menu.setMenuProducts(Collections.singletonList(MenuProductTestFixture.create().getMenuProduct()));
        return new MenuTestFixture(menu);
    }

    public MenuTestFixture changeId(UUID id) {
        Menu newMenu = new Menu();
        newMenu.setId(id);
        newMenu.setName(menu.getName());
        newMenu.setPrice(menu.getPrice());
        newMenu.setMenuGroup(menu.getMenuGroup());
        newMenu.setMenuGroupId(menu.getMenuGroupId());
        newMenu.setDisplayed(menu.isDisplayed());
        newMenu.setMenuProducts(menu.getMenuProducts());
        this.menu = newMenu;
        return this;
    }

    public Menu getMenu() {
        return this.menu;
    }
}
```

이 방법이 구현 코드가 많긴한데 한 번 구현해놓으면 편하더라구요 ㅎㅎ.. 
Builder 와 거의 유사하다고 보시면 됩니다
이게 결코 정답은 아니니 시창님의 입맛에 맞게 구현하시면 됩니다!!

jivebreaddev replied:
제가 질문이 부정확했는데 정확하게 알아주셔서 감사합니다!

Builder를 생성해두고 필요한 상태만 변경할 수 있는 것으로 변경하는것이군요!

요긴하게 사용하겠습니다! 감사합니다!

jivebreaddev replied:
![image](https://github.com/next-step/ddd-legacy/assets/89639413/c4a9ee73-4271-4f06-be2f-1f43268c0a19)


이런 경우도 order에 상태값만 바꿔서 바로 테스트 해볼수 있겠네요!

이번에는 SERVICE 로직 태워서 진행해보겠습니다!

Gomding commented on src/test/java/kitchenpos/application/menu/MenuServiceTest.java:
```diff
@@ -0,0 +1,201 @@
+  private FakeUuidBuilder fakeUuidBuilder;
+
+  private Product udon;
+  private Product ramen;
```
요건 사용하지 않으니 제거해도 되겠군요!

Gomding commented on src/test/java/kitchenpos/application/order/OrderServiceTest.java:
```diff
@@ -0,0 +1,275 @@
+  class orderAcceptance {
+
+    @Test
+    @DisplayName("주문의 상태를 `접수`로 변경한다.")
```
접수로 변경할 수 없는 케이스도 테스트에 있으면 좋겠군요!

접수로 변경할 수 없는 상태도 있을것 같아서요 :)

jivebreaddev replied:
생성해뒀습니다!

Gomding commented on src/test/java/kitchenpos/application/order/OrderServiceTest.java:
```diff
@@ -0,0 +1,275 @@
+
+    @DisplayName("주문의 상태가 `접수`이어야 한다.")
+    @Test
+    void passOrderWithOrderStatusAccepted() {
```
여기도 다른 상태(대기중, 배달중, 배달완료, 주문완료) 일때 변경할 수 없다는 것을 검증하지 않아도 괜찮을까요!

jivebreaddev replied:
생성해뒀습니다!

Gomding reviewed (changes requested):
안녕하세요 시창님!
변경사항 push 가 안된것 같아
해당 부분제외하고 추가 코멘트 남겼어요 :)

궁금한 점 있으면 편하게 DM이나 코멘트 남겨주세요 🤗

Jivebread committed:
test(Menu, Order, OrderTable ServiceTest): test 내의 사용하지 않는 fields 제거

Jivebread committed:
test(Menu, Order, OrderTable ServiceTest): test 내의 패키지 구조 변경

Jivebread committed:
test(Menu, Order, ServiceTest): 주문 상태의 값이 올바르지 않을때의 테스트 추가

Jivebread committed:
test(MenuCategoryNameTest): 메뉴 카테고리 생성에 대한 테스트인데 `vo에 대한 테스트`로 혼용

테스트 대상 Service 로직은 메뉴 카테고리 생성
해당 테스트는 MenuCategoryName 생성에 대한 테스트 로 표기됨
정상적인 이름일 때 메뉴 카테고리를 생성할 수 있다 로 변경이 필요함

jivebreaddev commented:
![image](https://github.com/next-step/ddd-legacy/assets/89639413/e3b4e42f-9535-40e6-ae98-87f7ac31cdc4)
## 요구사항
1. 사용하지 않는 FIELDS들 제거
2. TEST 내 패키지 구조 변경
3. 주문 상태가 올바르지 않을때 주문상태 변경에 대한 테스트 추가
4. 혼란스러운 TEST 이름에 대한 수정


추가적으로 피드백 주실 부분있으면 언제든지 환영입니다!!

Gomding commented on src/test/java/kitchenpos/application/menu/MenuServiceTest.java:
```diff
@@ -82,32 +75,36 @@ void createMenu() {
     @ParameterizedTest
     @ValueSource(strings = {""})
     void failToCreateMenuWithEmptyString(final String name) {
+      Product productUdon = productRepository.save(ProductFixture.createProduct(UDON, TWENTY_THOUSANDS, fakeUuidBuilder));
```
헉 이 부분은 필드 제거하고 테스트 내부로 넣으라는 의도는 아니었는데
필드를 제거해주셨군요 🤗

>   private Product ramen;

요 필드는 사용하는곳이 없길래 드렸던 코멘트였습니다 :)

혹은 다른 의도로 변경하신거라면 말씀해주세요..!

Gomding commented on src/test/java/kitchenpos/application/order/OrderTableServiceTest.java:
```diff
@@ -23,11 +23,11 @@
+import kitchenpos.infra.menu.FakeMenuRepository;
+import kitchenpos.infra.order.FakeOrderRepository;
+import kitchenpos.infra.order.FakeOrderTableRepository;
+import kitchenpos.infra.product.FakeProductRepository;
```
테스트 내부 패키지까지 꼼꼼하게 변경해주셨네요 👍 👍

Gomding commented on src/test/java/kitchenpos/application/order/OrderServiceTest.java:
```diff
@@ -165,21 +175,21 @@ void passOrderWithOrderStatusAccepted() {
-    @Test
-    void deliverOrderWithOrderStatusDelivering() {
+    @ParameterizedTest
+    @EnumSource(mode = INCLUDE, names = {"WAITING", "ACCEPTED", "SERVED", "DELIVERED", "COMPLETED"})
```
크 EnumSource 까지 활용해주셨군요 👍

Gomding reviewed (approved):
안녕하세요 시창님!
처음부터 테스트 꼼꼼하게 작성해주셔서 변경사항이 많지 않았네요 :)
요구사항은 모두 만족하셔서 이번 미션은 merge 하겠습니다!!
추가로 궁금한 부분있으시면 DM으로 의견나눠봐도 좋아요 :)
고생하셨습니다 🙌
남은 미션도 화이팅입니다!!
